### ISSUE TITLE
Composing a depolarizing error with a thermal error fails for certain values

### ISSUE THREAD

**Issue by bjader** [Qiskit/qiskit-aer#176]
### Informations

- **Qiskit Aer version**: 0.1.1
- **Python version**:  3.6.8
- **Operating system**: Ubuntu 18.04.2

### What is the current behavior?
When using the function `aer.noise.errors.quantum_error.compose` to combine a depolarizing error with a thermal error, an error is raised:
```
File "anaconda3/envs/Qiskitenv/lib/python3.6/site-packages/qiskit/providers/aer/noise/errors/quantum_error.py", line 234, in compose
    return QuantumError(noise_ops)
  File "/home/jaderberg/anaconda3/envs/Qiskitenv/lib/python3.6/site-packages/qiskit/providers/aer/noise/errors/quantum_error.py", line 102, in __init__
    raise NoiseError("Probabilities are not normalized: {} != 1".format(total_probs))
qiskit.providers.aer.noise.noiseerror.NoiseError: 'Probabilities are not normalized: 0.999999999875001 != 1'
```
### Steps to reproduce the problem
Run the following code:
```
from qiskit import IBMQ
from qiskit.providers.aer.noise import NoiseModel
from qiskit.providers.aer.noise.errors import pauli_error, thermal_relaxation_error, depolarizing_error
noisemodel = NoiseModel()
t1 = 3000 * 1000
t2 = 3000 * 1000
gate_time = 50
error_thermal = thermal_relaxation_error(t1, t2, gate_time)
p_depol = 1e-5
error_depol = depolarizing_error(p_depol, 1)
error_depol.compose(error_thermal)
```
### What is the expected behaviour?
The compose function should combine the errors successfully, as is the case if you change the value of `p_depol` in the previous example to 1e-4.

### Suggested solutions
`compose` initialises a `QuantumError` object with the default threshold value of 1e-10. However, the noise circuits it creates have a combined probability further away from 1 than this default threshold. This causes an error to be raised.

Possibilities and open questions:
1. Where did the 1e-10 default threshold come from in the __init__ for `QuantumError`. Is it sufficient?
1. Should the `compose` function use the default threshold, or have some higher tolerance?
1. Why does the `compose` function create noise circuits with a total probability that is further away from 1 than the default threshold?

**Comment by chriseclectic** [comment]
I think this should be fixed in master. The default absolute tolerance parameter was relaxed to 1e-8, and you can also change this value yourself with `error.atol = val`

**Comment by bjader** [comment]
Hi @chriseclectic thanks for responding to this. How would I go about using the master version as my conda environment? Also, do you know when the next release will be which would have this change in it?

**Comment by bjader** [comment]
Having relaxed the the absolute tolerance parameter to 1e-8 manually I still have a huge range of values where compose the same error.

Below are two graphs showing the infidelity of the cnot gate based upon input parameters to a simple noise model consisting of thermal relaxation error and depolarizing errors. My noise model is based largely from this link https://github.com/Qiskit/qiskit-presentations/blob/master/2019-02-26_QiskitCamp/QiskitCamp_Simulation.ipynb.

Figure 1
![image](https://user-images.githubusercontent.com/14994219/56906031-5fb8d500-6a99-11e9-9487-92981fbd2344.png)

Figure 2
![image](https://user-images.githubusercontent.com/14994219/56906051-66474c80-6a99-11e9-9d98-2ee13f7daae2.png)

Figure 1 is for a QuantumError.atol = 1e-10 and Figure 2 is for a QuantumError.atol = 1e-8. The important feature in both figures is the light green band, where an error was thrown and so the infidelity set to 1. What I'm trying to say is that whilst my initial example is fixed by change from 1e-10 to 1e-8, there are still a whole other range of values where you cannot construct a noise model. 

It is only when I reduce the tolerance all the way to 1e-2 does this behaviour disappear:

![image](https://user-images.githubusercontent.com/14994219/56906178-9f7fbc80-6a99-11e9-8cf0-99f32ff1d3f8.png)

**tldr:** Figure 2 shows a whole range of possible p_depol and t1/t2 times where my initial error can reproduced even with the absolute tolerance set to 1e-8. Just run the following code:

```python
from qiskit import IBMQ
from qiskit.providers.aer.noise import NoiseModel
from qiskit.providers.aer.noise.errors import pauli_error, thermal_relaxation_error, depolarizing_error
noisemodel = NoiseModel()
t1 = # Pick from light green region in Figure 2 and convert to nanoseconds
t2 = t1
gate_time = 200
error_thermal = thermal_relaxation_error(t1, t2, gate_time)
p_depol = 1e-5
error_depol = depolarizing_error(p_depol, 1)
error_depol.compose(error_thermal)
```

**Comment by bjader** [comment]
For purposes of reproducing my issue, here is the full noise model I used to construct these figures.

```python
class NoiseModelBuilder:
    """
    A class to help build a custom NoiseModel from scratch. Many of the functions are based on examples from
    https://github.com/Qiskit/qiskit-presentations/blob/master/2019-02-26_QiskitCamp/QiskitCamp_Simulation.ipynb
    """

    def __init__(self, coupling_map=None):
        self.noise_model = NoiseModel()
        self.coupling_map = coupling_map

    def construct_basic_device_model(self, p_depol1, p_depol2, t1, t2):
        """
        A noise model to emulate qiskit.providers.aer.noise.device.models.basic_device_noise_model().

        The noise model includes the following errors:

            * Single qubit readout errors on measurements.
            * Single-qubit gate errors consisting of a depolarizing error
              followed by a thermal relaxation error for the qubit the gate
              acts on.
            * Two-qubit gate errors consisting of a 2-qubit depolarizing
              error followed by single qubit thermal relaxation errors for
              all qubits participating in the gate.

        :param p_depol1: Probability of a depolarising error on single qubit gates
        :param p_depol2: Probability of a depolarising error on two qubit gates
        :param t1: Thermal relaxation time constant
        :param t2: Dephasing time constant
        """
        logging.info(
            "Creating basic noise model with p_depol1 = {0}, p_depol2 = {1}, t1 = {2}, t2 = {3}.".format(p_depol1,
                                                                                                         p_depol2, t1,
                                                                                                         t2))

        if t2 > 2 * t1:
            raise ValueError("t2 cannot be greater than 2t1")

        # Thermal relaxation error
        # Instruction times (in nanoseconds)
        time_u1 = 0  # virtual gate
        time_u2 = 50  # (single X90 pulse)
        time_u3 = 100  # (two X90 pulses)
        time_cx = 300
        time_reset = 1000  # 1 microsecond
        time_measure = 1000  # 1 microsecond

        # QuantumError objects
        error_reset = thermal_relaxation_error(t1, t2, time_reset)
        error_measure = thermal_relaxation_error(t1, t2, time_measure)
        error_thermal_u1 = thermal_relaxation_error(t1, t2, time_u1)
        error_thermal_u2 = thermal_relaxation_error(t1, t2, time_u2)
        error_thermal_u3 = thermal_relaxation_error(t1, t2, time_u3)
        error_thermal_cx = thermal_relaxation_error(t1, t2, time_cx).kron(thermal_relaxation_error(t1, t2, time_cx))

        # Depolarizing error
        error_depol1 = depolarizing_error(p_depol1, 1)
        error_depol2 = depolarizing_error(p_depol2, 2)

        self.noise_model.add_all_qubit_quantum_error(error_reset, "reset")
        self.noise_model.add_all_qubit_quantum_error(error_measure, "measure")

        self.noise_model.add_all_qubit_quantum_error(error_depol1.compose(error_thermal_u1), "u1")
        self.noise_model.add_all_qubit_quantum_error(error_depol1.compose(error_thermal_u2), "u2")
        self.noise_model.add_all_qubit_quantum_error(error_depol1.compose(error_thermal_u3), "u3")

        for pair in self.coupling_map:
            self.noise_model.add_quantum_error(error_depol2.compose(error_thermal_cx), "cx", pair)
```

**Comment by chriseclectic** [comment]
Thanks @bjader, there might be something going on with an accumulation of rounding errors in the conversions from superoperators to Kraus matrices that needs to be regularized. I'll look into it. Could you post the code you used to make the plot?

**Comment by bjader** [comment]
Hi @chriseclectic thanks for taking the time to read that above - that definitely sounds like something that could be going on. Here's the plotting code, which also uses the `NoiseModelBuilder` class I posted above:

```python
import logging

import numpy as np
from matplotlib import pyplot as plt
from matplotlib.colors import LogNorm
from qiskit.providers.aer.noise.noiseerror import NoiseError

from classicalmethods.noisemodel_builder import NoiseModelBuilder

coupling_map = [[0, 1], [0, 2], [0, 3], [0, 4], [1, 0], [1, 2], [1, 3], [1, 4], [2, 0], [2, 1], [2, 3], [2, 4], [3, 0],
                [3, 1], [3, 2], [3, 4], [4, 0], [4, 1], [4, 2], [4, 3]]

tau_list = np.array((range(50, 10001, 50)))  # 50us to 10,000us in steps of 50

j = np.array(range(4, 40))
exponent = j / 4.0
p_depol_list = 10 ** (-exponent)  # 10^-1 to 10^-10 in steps of 10^0.25
cx_fidelity_list = []

for tau in tau_list:
    print("tau = {}".format(tau))
    cx_fixed_tau = []
    for p_depol in p_depol_list:
        builder = NoiseModelBuilder(coupling_map)
        try:
            builder.construct_basic_device_model(p_depol, p_depol, tau * 1000, tau * 1000)
            noise_model = builder.noise_model

            pairs = noise_model._local_quantum_errors.get('cx')
            error = list(pairs.values())[0]
            cx_fidelity = max(error[0]._noise_probabilities)

        except NoiseError as e:
            logging.warning("Error occurred for tau = {0}, p_depol = {1}".format(tau, p_depol))
            cx_fidelity = 0

        cx_fixed_tau.append(cx_fidelity)
    cx_fidelity_list.append(cx_fixed_tau)

fidelity = np.array(cx_fidelity_list)
infidelity = 1 - fidelity

plt.xlabel(r'$p_{depol}$')
plt.xscale('log')
plt.gca().invert_xaxis()

plt.ylabel(r'Qubit lifetime $\mu s$')

levels = np.logspace(-5, 0, 6)
plt.contourf(p_depol_list, tau_list, infidelity, levels=levels, norm=LogNorm(infidelity.min(), infidelity.max()))
colorbar = plt.colorbar()
colorbar.set_label("Infidelity")

plt.title("Infidelity of cx gate based on noise model parameters")
plt.show()
```

**Comment by chriseclectic** [comment]
I think the problem was that the atol value was also used for truncating probabilities, which meant increasing the atol value could make the problem worse. I have fixed that in #175, and also added a renormalization of probabilities that pass a sum check against the atol value just to be safe. Running your code on that branch with default atol I get:

![image](https://user-images.githubusercontent.com/2235104/56914822-165d8b00-6a83-11e9-90d9-5b9312b52317.png)

**Comment by bjader** [comment]
Hi @chriseclectic that's fantastic, thanks for fixing this! Assuming that it may be a while until the next release, what would be the best way for me to take on this change from master now? Baring in mind that my current setup is using a conda virtualenv.
